File: leetcode/medium/129-sum-root-to-leaf-numbers/main.py
# 129. Sum Root to Leaf Numbers
# Given a binary tree containing digits from 0-9 only, each root-to-leaf path could represent a number.

# An example is the root-to-leaf path 1->2->3 which represents the number 123.
# Find the total sum of all root-to-leaf numbers.
# Note: A leaf is a node with no children.


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        # return self.dfs(root, 0)
        return self.bfs(root)
    
    # Time complexity: O(n), Space complexity: O(d)
    def dfs(self, root: TreeNode, sum: int) -> int:
        if not root:
            return 0
        newSum = sum * 10 + root.val
        
        isLeaf = not root.left and not root.right
        if isLeaf:
            return newSum
        
        return self.findLeafSum(root.left, newSum) + self.findLeafSum(root.right, newSum)
    
    # Time complexity: O(n), Space complexity: O(n)
    def bfs(self, root: TreeNode) -> int:
        if not root:
            return 0
        allSum = 0
        queue = [(root, 0)]
        
        while queue:
            node, sum = queue.pop()
            
            if not node.left and not node.right:
                allSum += sum * 10 + node.val
            
            if node.right:
                queue.insert(0, (node.right, sum * 10 + node.val))
            
            if node.left:
                queue.insert(0, (node.left, sum * 10 + node.val))
        
        return allSum

File: leetcode/medium/129-sum-root-to-leaf-numbers/main.swift
/**
    129. Sum Root to Leaf Numbers

    Given a binary tree containing digits from 0-9 only, each root-to-leaf path could represent a number.
    An example is the root-to-leaf path 1->2->3 which represents the number 123.
    Find the total sum of all root-to-leaf numbers.

    Note: A leaf is a node with no children.
        
    Time complexity: O(n)
    Space complexity: O(n)
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func sumNumbers(_ root: TreeNode?) -> Int {
        var result = 0
        dfs(root, 0, &result)
        return result
    }
    
    private func dfs(_ root: TreeNode?, _ current: Int, _ result: inout Int) {
        guard let root = root else { return }
        
        let newCurrent = current * 10 + root.val

        let isLeaf = root.left == nil && root.right == nil
        if isLeaf {
            
            result += newCurrent
            return
        }
        
        dfs(root.left, newCurrent, &result)
        dfs(root.right, newCurrent, &result)
    }
}
